hax/hax/motr/planner.py

```diff
@@ -0,0 +1,207 @@
+            # notify them
+            self.b_lock.notifyAll()
+
+    def _create_poison(self) -> BaseMessage:
+        cmd = Die()
+        cmd.group = self.state.current_group_id
+        return cmd
+
+    def get_next_command(self) -> BaseMessage:
+        while True:
+            LOG.log(TRACE, '[WP]Trying to get new command')
+            with self.b_lock:
+                if self.state.is_shutdown:
+                    return self._create_poison()
+                if self.backlog:
+                    cmd = self.backlog.popleft()
+                    LOG.log(TRACE, '[WP]Cmd %s taken!', cmd)
+                    self.state.taken_commands.add(cmd)
+                    return cmd
+                LOG.log(TRACE, '[WP]Blocking thread: no commands in backlog')
+                self.b_lock.wait()
+
+    def shutdown(self):
+        with self.b_lock:
+            LOG.debug('WorkPlanner is shutting down')
+            self.state.is_shutdown = True
+            self.b_lock.notifyAll()
+
+    def ensure_allowed(self, command: BaseMessage) -> BaseMessage:
+        def is_current(cmd: BaseMessage, st: State) -> bool:
+            return cmd.group == st.current_group_id
+
+        while True:
+            with self.b_lock:
+                state = self.state
+                if state.is_shutdown:
+                    return self._create_poison()
+                if is_current(command, state):
+                    state.taken_commands.remove(command)
+                    state.active_commands.add(command)
+                    return command
+                LOG.log(TRACE, '[WP]Cmd %s not allowed for now.'
+                        ' Current state: %s', command, self.state)
+                self.b_lock.wait()
+
+    def _inc_group(self):
+        state = self.state
+        state.current_group_id += 1
+        if state.next_group_id < state.current_group_id:
+            state.next_group_id = state.current_group_id
+            state.next_group_commands = LinkedSet()
+
+    def notify_finished(self, command: BaseMessage) -> None:
+        with self.b_lock:
+            state = self.state
+            state.active_commands.remove(command)
+            LOG.log(TRACE, '[WP]Cmd %s removed. Current state: %s', command,
+                    state)
+
+            if state.active_commands:
+                return
+            for c in self.backlog:
+                if c.group == state.current_group_id:
+                    return
+            for c in state.taken_commands:
+                if c.group == state.current_group_id:
+                    return
+            # if we're here, command was the only one belonging to group
+            self._inc_group()
+            LOG.log(TRACE, '[WP]Active group changed to %s',
+                    state.current_group_id)
+            # The group changed, let's unblock those who are waiting for
+            # this group
+            self.b_lock.notifyAll()
+
+    def _should_increase_group(self, cmd: BaseMessage) -> bool:
```

Comment:
How about fix set of groups with an allowed_message bitmap? We have fixed set of commands, we can have a fixed set of groups where each group takes only those messages (commands) which are allowed by the allowed_message bitmap in the group. This can be more efficient than searching in the list and we can also check the invariant easily.

Reply:
I'll check if standard pythonic `set` can hold types. That would be simpler than maintaining bitmap, map the types to integer constants etc.

Reply:
Addressed. Standard `set()` is used for types given they are hashable.